flm serve --asr 1 doesn't raise the memlock limit as expected

I'm trying to run flm without having memlock set to infinity. This is important for lemonade which only passes Linux capabilities to the flm process.

To replicate this behaviour manually, first unset the memlock limits. Then use the following command:

`sudo setcap CAP_SYS_RESOURCE+ep /usr/bin/flm` to add the capability to the binary,

`sudo setcap CAP_SYS_RESOURCE-ep /usr/bin/flm` to remove the capability from the binary

When I try to run whisper or embed standalone, I get an error:

```
gso@server ~[1]> flm serve --asr 1
[FLM]  Using environment-specified port: 52625
Error: mmap(addr=0x7f8218000000, len=67108864, prot=3, flags=8209, offset=4294967296) failed (err=-11): Resource temporarily unavailable

```

If I add an llm model like gemma3:4b, it raises the memlock limit and runs normally.
```
gso@server ~ [1]> flm serve gemma3:4b --asr 1
[Linux]  Memlock Limit: raised to 4326 MB for gemma3:4b
[FLM]  Using environment-specified port: 52625
[FLM]  Loading model: /home/gso/.config/flm/models/Whisper-V3-Turbo-NPU2
[FLM]  Loading model: /home/gso/.config/flm/models/Gemma3-4B-NPU2
[FLM]  Starting server on port 52625...
[FLM]  WebServer started on port 52625 with 10 I/O threads
```

> [@sofiageo](https://github.com/sofiageo) Could you please try the artifacts [here](https://github.com/FastFlowLM/FastFlowLM/actions/runs/24851177679) to see if the issue is solved?

Sorry I can't run the artifact. I'm on Arch Linux and while I did try many things it didn't work for me. But I can build the PR fine. This is what I get when I run the PR.

Running validate **without** capabilities:
```
gso@server ~> flm validate
[ERROR]  Memlock limit is too low (8MB). Please raise the limit or set to infinity.
```
Running embed **without** capabilities:
```
gso@server ~> flm serve --embed 1
[FLM]  Embedding mode enabled: reserving additional 300MB of memory
[Linux]  Warning: could not raise memlock limit to 798 MB
Error: mmap(addr=0x7f0e88000000, len=67108864, prot=3, flags=8209, offset=4294967296) failed (err=-11): Resource temporarily unavailable
```
Running asr **without** capabilities:
```
gso@server ~> flm serve --asr 1
[FLM]  ASR mode enabled: reserving additional 1GB of memory
[Linux]  Warning: could not raise memlock limit to 1465 MB
Error: mmap(addr=0x7f3e60000000, len=67108864, prot=3, flags=8209, offset=4294967296) failed (err=-11): Resource temporarily unavailable
```
Running embed **with** capabilities:
```
gso@server ~> sudo setcap CAP_SYS_RESOURCE+ep /usr/bin/flm
gso@server ~> flm serve --embed 1
[FLM]  Embedding mode enabled: reserving additional 300MB of memory
[Linux]  Memlock Limit: raised to 798 MB
```
Running asr **with** capabilities:
```                                                                                                                                                                                                                                                                                       gso@server ~> flm serve --asr 1
[FLM]  ASR mode enabled: reserving additional 1GB of memory
[Linux]  Memlock Limit: raised to 1465 MB
```

I've also tested it with lemonade and it's working fine. Thanks for the fix! I like how you solved it without making it more complicated that it needs to be 💯

--- a/src/src/main.cpp
+++ b/src/src/main.cpp
@@ -529,22 +529,36 @@ int main(int argc, char* argv[]) {
 
 #ifndef _WIN32
     // Raise memlock limit to accommodate the model being loaded
-    if ((parsed_args.command == "run" || parsed_args.command == "serve" || parsed_args.command == "bench") &&
-        parsed_args.model_tag != "model-faker" && !parsed_args.model_tag.empty()) {
-        auto [resolved_tag, model_info] = availble_models.get_model_info(parsed_args.model_tag);
-        if (model_info.contains("size")) {
-            rlim_t model_size = model_info["size"].get<uint64_t>();
-            // Add 512MB overhead for working memory beyond the model itself
-            rlim_t required = model_size + (512ULL * 1024 * 1024);
-            struct rlimit rl;
-            if (getrlimit(RLIMIT_MEMLOCK, &rl) == 0) {
-                if (rl.rlim_cur != RLIM_INFINITY && rl.rlim_cur < required) {
-                    struct rlimit rl_new = { required, required };
-                    if (setrlimit(RLIMIT_MEMLOCK, &rl_new) == 0) {
-                        header_print_g("Linux", "Memlock Limit: raised to " << (required / 1024 / 1024) << " MB for " << resolved_tag);
-                    } else {
-                        header_print("Linux", "Warning: could not raise memlock limit to " << (required / 1024 / 1024) << " MB");
-                    }
+    if ((parsed_args.command == "run" || parsed_args.command == "serve" || parsed_args.command == "bench")) {
+        rlim_t model_size = 0;
+        rlim_t asr_size = 0;
+        rlim_t embedding_size = 0;
+
+        if (parsed_args.model_tag != "model-faker" && !parsed_args.model_tag.empty()) {
+            auto [resolved_tag, model_info] = availble_models.get_model_info(parsed_args.model_tag);
+            if (model_info.contains("size")) {
+                model_size = model_info["size"].get<uint64_t>();
+            }
+        }
+        if (parsed_args.asr) {
+            asr_size = 1000000000;
+            header_print_g("FLM", "ASR mode enabled: reserving additional 1GB of memory");
+        }
+        if (parsed_args.embed) {
+            embedding_size = 300000000;
+            header_print_g("FLM", "Embedding mode enabled: reserving additional 300MB of memory");
+        }
+
+        // Add 512MB overhead for working memory beyond the model itself
+        rlim_t required = model_size + asr_size + embedding_size + (512ULL * 1024 * 1024);
+        struct rlimit rl;
+        if (getrlimit(RLIMIT_MEMLOCK, &rl) == 0) {
+            if (rl.rlim_cur != RLIM_INFINITY && rl.rlim_cur < required) {
+                struct rlimit rl_new = { required, required };
+                if (setrlimit(RLIMIT_MEMLOCK, &rl_new) == 0) {
+                    header_print_g("Linux", "Memlock Limit: raised to " << (required / 1024 / 1024) << " MB");
+                } else {
+                    header_print("Linux", "Warning: could not raise memlock limit to " << (required / 1024 / 1024) << " MB");
                 }
             }
         }

